BUG: Fix rounding of denormals in double and float to half casts  …

seberg opened the pull request:
Previously the significand was shifted right to align denormals of
different magnitude. This loses some bits that can make a difference
for rounding. This is fixed:
 1. For floats, by inspecting the original last bits when this may
    make a difference (should happen rarely)
 2. For doubles by shifting the bits left to align the denromals and
    thus not lose the lowest orginal bits.

-----

I am not sure if the double branch is not being a bit too smart and it is nicer to just also use the `||` logic from above, since it uses different constants compared to the normal branch because of this.

I do not know if there may be nicer ways with larger code changes, but I could not think of one.

This closes gh-12721

seberg committed:
BUG: Fix rounding of denormals in double and float to half casts

Previously the significand was shifted right to align denormals of
different magnitude. This loses some bits that can make a difference
for rounding. This is fixed:
 1. For floats, by inspecting the original last bits when this may
    make a difference (should happen rarely)
 2. For doubles by shifting the bits left to align the denromals and
    thus not lose the lowest orginal bits.

seberg committed:
TST: Test half denormal rounding

The test assumes that half to even is active, if this is ever
changed or allowed to change, the test will fail.

mhvk commented on numpy/core/tests/test_half.py:
```diff
@@ -69,6 +69,58 @@ def test_half_conversions(self):
 
+    @pytest.mark.parametrize("offset", [None, "up", "down"])
+    @pytest.mark.parametrize("shift", [None, "up", "down"])
+    @pytest.mark.parametrize("float_t", [np.float32,])
```
Also test with `np.float64`? Otherwise why parametrize it?

seberg replied:
Yeah, forgot to read after trying, hopefully the float64s fail ;) because of that 3 bit thing above.

mhvk reviewed (commented):
One possibly significant comment about the changes.

If this is wrong, then we need more tests, since the tests we have all passed...

eric-wieser commented on numpy/core/src/npymath/halffloat.c:
```diff
@@ -307,9 +307,11 @@ npy_uint16 npy_floatbits_to_halfbits(npy_uint32 f)
+         * In all other cases, we can just add one.
          */
-        if ((f_sig&0x00003fffu) != 0x00001000u) {
+        if (((f_sig&0x00003fffu) != 0x00001000u) || (f&0x008ffu)) {
```
Where did `008ff` come from? And why are you giving only 5 hex digits?

seberg replied:
Jeesh, I messed up, probably a typo, it should have been `7ff` and of course full digits. 7ff for the last 11 bits, that would otherwise be gone.

seberg replied:
Dang, the tests probably do not catch this, because it is in the wrong bits... maybe should add a test for this part more specifically.

eric-wieser commented on numpy/core/src/npymath/halffloat.c:
```diff
@@ -416,21 +418,26 @@ npy_uint16 npy_doublebits_to_halfbits(npy_uint64 d)
          */
-        if ((d_sig&0x000007ffffffffffULL) != 0x0000020000000000ULL) {
-            d_sig += 0x0000020000000000ULL;
+        if ((d_sig&0x003fffffffffffffULL) != 0x0010000000000000ULL) {
```
Might be more readable to express these as `(1 << n) - 1` and `(1 << n)` for some `n`. Fine to leave as it too.

seberg replied:
Yeah, should change it everywhere then though probably.

seberg replied:
Probably a good idea, these bit patterns are a error prone :/, so I may do that but not super soon maybe. For now I had already fixed it up and added a test for the one actual problem.

mattip replied:
spaces needed around `&`

eric-wieser replied:
Without spaces is par for the course for this file - I don't think we need to block this PR on that.

seberg commented on numpy/core/src/npymath/halffloat.c:
```diff
@@ -416,21 +418,26 @@ npy_uint16 npy_doublebits_to_halfbits(npy_uint64 d)
         }
 #else
-        d_sig += 0x0000020000000000ULL;
+        d_sig += 0x0000e00000000000ULL;
```
wow, yes you are right :(, that is embaressing.

seberg replied:
Ah, that was some left-over in a code path that is never used (and thus cannot be tested), we could remove that completely in principle.

mhvk replied:
Yes, I guess we always do proper round-to-even now.

seberg committed:
Fixup: Fixup for halffloat.c

The one code path cannot be used. The other must have been a typo, but
is a valid bug, a new test for it in the next commit.

seberg committed:
TST: half casting lower bits are not lost for denormal results

The first test only tested the off by one, this one specifically tests
that all bits are used to decide if "round to nearest even" should
be used, in the example of rounding towards 0.

seberg committed:
TST: Test not just denormals but all positive finite float16s

eric-wieser commented:
Would having #10741 make testing easier here?

eric-wieser commented on numpy/core/src/npymath/halffloat.c:
```diff
@@ -416,21 +418,26 @@ npy_uint16 npy_doublebits_to_halfbits(npy_uint64 d)
+         * significand. This changes the constant compared to normal branch.
+         */
+        assert(d_exp - 998 >= 0);
+        d_sig <<= (d_exp - 998);
```
If you can shed any light in a comment on on where `998` comes from, that would be appreciated.

mattip commented:
LGTM. Still needs release note

seberg committed:
DOC: Add lots of comments and add a short release note.

seberg commented:
Hmm, I thought/think release notes make sense, because it is a subtle change (we can still remove it anyway). But would do we want to backport it, and so that it should be already in 1.16.1 or both?

seberg commented:
Plus, in that case I should maybe look into giving it a nice backportable merge base.

charris commented:
> But would do we want to backport it, and so that it should be already in 1.16.1 or both?

Needs to be decided for the LTS branch, we don't yet have a policy for that.  I'm inclined to go for duplicate entries because the LTS branch is the end of Python 2.7 support, so a bit more separate than usual.

seberg commented:
Sorry, got a bit side tracked, I don't quite remember myself if I thought it was. I think I was considering doing the simple fix in both versions still. I doubt it makes a noticable speed difference anyway.

charris commented:
OK, no more feedback, so putting this in. Thanks Sebastian.
